diff --git a/fftools/ffmpeg.c b/fftools/ffmpeg.c
index de607cac93..8b87093cf8 100644
--- a/fftools/ffmpeg.c
+++ b/fftools/ffmpeg.c
@@ -83,9 +83,21 @@
 #include "ffmpeg_utils.h"
 #include "graph/graphprint.h"
 
-const char program_name[] = "ffmpeg";
+const char program_name[] = "ffmpeg 2mmilab 15-06-2025";
 const int program_birth_year = 2000;
 
+static uint64_t total_size_bytes = 0;
+static int slow_delay = 0;
+static int slow = 0;
+
+void ff_add_input_bytes(int size)
+{
+    if (size > 0)
+    {
+        total_size_bytes += size;
+    }
+}
+
 FILE *vstats_file;
 
 typedef struct BenchmarkTimeStamps {
@@ -677,6 +689,9 @@ static void print_report(int is_last_report, int64_t timer_start, int64_t cur_ti
         av_bprintf(&buf_script, "speed=%4.3gx\n", speed);
     }
 
+    av_bprintf(&buf, " cursor=%" PRId64, total_size_bytes);
+    av_bprintf(&buf, " slow=%d", slow);
+
     secs = (int)t;
     ms = (int)((t - secs) * 1000);
     mins = secs / 60;
diff --git a/fftools/ffmpeg.h b/fftools/ffmpeg.h
index 7868f3d85f..87feac6580 100644
--- a/fftools/ffmpeg.h
+++ b/fftools/ffmpeg.h
@@ -931,4 +931,6 @@ int view_specifier_parse(const char **pspec, ViewSpecifier *vs);
 int muxer_thread(void *arg);
 int encoder_thread(void *arg);
 
-#endif /* FFTOOLS_FFMPEG_H */
+void ff_add_input_bytes(int size);
+
+#endif /* FFTOOLS_FFMPEG_H */
\ No newline at end of file
diff --git a/fftools/ffmpeg_dec.c b/fftools/ffmpeg_dec.c
index b06455c7b1..019b7fa92a 100644
--- a/fftools/ffmpeg_dec.c
+++ b/fftools/ffmpeg_dec.c
@@ -715,6 +715,11 @@ static int packet_decode(DecoderPriv *dp, AVPacket *pkt, AVFrame *frame)
     }
 
     if (pkt) {
+
+        if (pkt->size > 0) { // Ensure we only add positive sizes
+            ff_add_input_bytes(pkt->size); // Call the function from ffmpeg.c
+        }
+
         FrameData *fd = packet_data(pkt);
         if (!fd)
             return AVERROR(ENOMEM);
