diff --git a/fftools/ffmpeg.c b/fftools/ffmpeg.c
index de607cac93..5a79504bcf 100644
--- a/fftools/ffmpeg.c
+++ b/fftools/ffmpeg.c
@@ -83,9 +83,21 @@
 #include "ffmpeg_utils.h"
 #include "graph/graphprint.h"
 
-const char program_name[] = "ffmpeg";
+const char program_name[] = "ffmpeg 2mmilab 15-06-2025";
 const int program_birth_year = 2000;
 
+static uint64_t total_size_bytes = 0;
+int slow_delay = 0;
+int slow = 0;
+
+void ff_add_input_bytes(int size)
+{
+    if (size > 0)
+    {
+        total_size_bytes += size;
+    }
+}
+
 FILE *vstats_file;
 
 typedef struct BenchmarkTimeStamps {
@@ -677,6 +689,9 @@ static void print_report(int is_last_report, int64_t timer_start, int64_t cur_ti
         av_bprintf(&buf_script, "speed=%4.3gx\n", speed);
     }
 
+    av_bprintf(&buf, " cursor=%" PRId64, total_size_bytes);
+    av_bprintf(&buf, " slow=%d", slow);
+
     secs = (int)t;
     ms = (int)((t - secs) * 1000);
     mins = secs / 60;
@@ -863,6 +878,17 @@ static int check_keyboard_interaction(int64_t cur_time)
                         "s      Show QP histogram\n"
         );
     }
+
+    /*slow down sleep */
+    if (key == '1'){
+        slow_delay = 100000;
+        slow = 1;
+    }
+    if (key == '0') {
+        slow_delay = 0;
+        slow = 0;
+    }
+
     return 0;
 }
 
diff --git a/fftools/ffmpeg.h b/fftools/ffmpeg.h
index 7868f3d85f..2d5270dbfd 100644
--- a/fftools/ffmpeg.h
+++ b/fftools/ffmpeg.h
@@ -931,4 +931,9 @@ int view_specifier_parse(const char **pspec, ViewSpecifier *vs);
 int muxer_thread(void *arg);
 int encoder_thread(void *arg);
 
-#endif /* FFTOOLS_FFMPEG_H */
+
+extern int slow_delay;
+extern int slow;
+void ff_add_input_bytes(int size);
+
+#endif /* FFTOOLS_FFMPEG_H */
\ No newline at end of file
diff --git a/fftools/ffmpeg_dec.c b/fftools/ffmpeg_dec.c
index b06455c7b1..083f908caa 100644
--- a/fftools/ffmpeg_dec.c
+++ b/fftools/ffmpeg_dec.c
@@ -715,6 +715,11 @@ static int packet_decode(DecoderPriv *dp, AVPacket *pkt, AVFrame *frame)
     }
 
     if (pkt) {
+
+        if (pkt->size > 0) { // Ensure we only add positive sizes
+            ff_add_input_bytes(pkt->size); // Call the function from ffmpeg.c
+        }
+
         FrameData *fd = packet_data(pkt);
         if (!fd)
             return AVERROR(ENOMEM);
diff --git a/fftools/ffmpeg_demux.c b/fftools/ffmpeg_demux.c
index 6e01efc420..ed26d6e0ee 100644
--- a/fftools/ffmpeg_demux.c
+++ b/fftools/ffmpeg_demux.c
@@ -460,6 +460,11 @@ static int input_packet_process(Demuxer *d, AVPacket *pkt, unsigned *send_flags)
     FrameData *fd;
     int ret = 0;
 
+    /*slow down 1 = slow 0 = full speed*/
+    if (slow_delay > 0) {
+        av_usleep(slow_delay);
+    }
+
     fd = packet_data(pkt);
     if (!fd)
         return AVERROR(ENOMEM);
